src/meta/duplication/meta_duplication_service.cpp

```diff
@@ -701,6 +702,10 @@ void meta_duplication_service::recover_from_meta_state()
         _meta_svc->get_meta_storage()->get_children(
             get_duplication_path(*app),
             [this, app](bool node_exists, const std::vector<std::string> &dup_id_list) {
+                dsn::defer([this, app]() {
```

Comment:
The memory state will be updated [1] after get data from ZK successfully in line 721, why update it again?

1. https://github.com/apache/incubator-pegasus/blob/master/src/meta/duplication/meta_duplication_service.cpp#L787

Reply:
> The memory state will be updated [1] after get data from ZK successfully in line 721, why update it again?
> 
> 1. https://github.com/apache/incubator-pegasus/blob/master/src/meta/duplication/meta_duplication_service.cpp#L787

Because in function `recover_from_meta_state`, the vector `dup_id_list` may be empty.
Here is the detail of our online condition:
1. Delete the existing hot standby first
    A. Delete the corresponding dup from zk
    B. Set duplicating in meta_server memory to 0 (not saved in zk), duplicating in zk is still 1
2. meta_server restarts
  Read dup and app_info data from zk to memory. At this time, dup is empty, but duplicating in zk is still 1 (because dup is empty, it is impossible to delete the hot standby to make duplicating 0).
3. Meta passes the hot standby status to the replica through `on_config_sync`
4.  Replica GC checks that the dup map is empty (`min_confirmed_decree = invalid_decree`), but `_is_duplication_master = true`, so GC is not performed

Reply:
Thanks for the information. So this patch is aim to complete step 2? Why not complete the 1.B step?

Reply:
> Thanks for the information. So this patch is aim to complete step 2? Why not complete the 1.B step?

Both two ways can fix this problem. In my opinion, the ways like step 2 is easier to implement.

Reply:
Correctness and consistent is important. Update the `duplicating` status to ZK is not very difficult if all duplications removed.
Or, if the `duplicating` memory status can be recovered from `<app>/duplication` ZK path, just competely remove the `duplicating` status from ZK.